src/handlers/web3.ts

```diff
@@ -217,27 +225,37 @@ export const getCachedProviderForNetwork = (network: Network = Network.mainnet)
  * @return The provider for the network.
  */
 export const getProviderForNetwork = async (network: Network | string = Network.mainnet): Promise<StaticJsonRpcProvider> => {
-  if (isNetworkEnum(network) && networkProviders[network]) {
-    return networkProviders[network]!;
+  const isSupportedNetwork = isNetworkEnum(network);
+  const cachedProvider = isSupportedNetwork ? networkProviders.get(network) : undefined;
+
+  if (isSupportedNetwork && cachedProvider) {
+    return cachedProvider;
   }
 
-  if (!isNetworkEnum(network)) {
+  if (!isSupportedNetwork) {
     const provider = new StaticJsonRpcProvider(network, Network.mainnet);
-    networkProviders[Network.mainnet] = provider;
+    networkProviders.set(Network.mainnet, provider);
     return provider;
   } else {
-    const chainId = getNetworkObj(network).id;
-    const provider = new StaticJsonRpcProvider(getNetworkObj(network).rpc, chainId);
-    if (!networkProviders[network]) {
-      networkProviders[network] = provider;
-    }
-    await provider.ready;
+    const provider = new StaticJsonRpcProvider(getNetworkObj(network).rpc(), getNetworkObj(network).id);
+    networkProviders.set(network, provider);
     return provider;
   }
 };
```

Comment:
the refactor keeps the same behavior, but is it right? if the network is supported but not cached we return `undefined`?

Reply:
If the network is supported but not cached, it'll skip this block:

```typescript
if (isSupportedNetwork && cachedProvider) {
  return cachedProvider;
}
```

And then it'll hit this else block:

```typescript
if (!isSupportedNetwork) {
  // ...
else {
  const provider = new StaticJsonRpcProvider(getNetworkObj(network).rpc(), getNetworkObj(network).id);
  networkProviders.set(network, provider);
  return provider;
}
```

Which creates a new provider and adds it to the cache — where can it return `undefined`?